test/integration/filters/header_to_proxy_filter.cc

```diff
@@ -0,0 +1,54 @@
+#include <string>
+
+#include "envoy/http/filter.h"
+#include "envoy/registry/registry.h"
+#include "envoy/server/filter_config.h"
+
+#include "source/common/network/filter_state_proxy_info.h"
+#include "source/common/network/utility.h"
+#include "source/extensions/filters/http/common/pass_through_filter.h"
+
+#include "test/extensions/filters/http/common/empty_http_filter_config.h"
+
+namespace Envoy {
+
+// A filter that looks for a specific header, and takes the address from that
+// header and inserts proxy override metadata.
+class HeaderToProxyFilter : public Http::PassThroughFilter {
+public:
+  Http::FilterHeadersStatus decodeHeaders(Http::RequestHeaderMap& request_headers, bool) override {
+    auto connect_proxy = Http::LowerCaseString("connect-proxy");
+    auto hostname = request_headers.getHostValue();
+    ASSERT(!hostname.empty());
+    if (!request_headers.get(connect_proxy).empty()) {
+      std::string address_string(request_headers.get(connect_proxy)[0]->value().getStringView());
+      auto address = Network::Utility::parseInternetAddressAndPort(address_string);
+      decoder_callbacks_->streamInfo().filterState()->setData(
```

Comment:
I wonder if we could use `LbEndpoint` itself somehow to determine the proxy address. In our case, each endpoint has an assigned proxy, and the proxy address is inside the endpoint metadata. I'm not sure how we'd make it work with a downstream filter.